tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/Unroller.scala

```diff
@@ -157,31 +158,64 @@ class Unroller(nameGenerator: UniqueNameGenerator, tracker: TransformationTracke
       decl: TlaDecl, bodyMap: BodyMap, inliner: InlinerOfUserOper, replaceWithDefaultsTr: TlaExTransformation
   ): TlaDecl = decl match {
     case d @ TlaOperDecl(name, fparams, body) if d.isRecursive =>
+      // In the case of recursively defined operators inside LET-IN, renaming has mangled their names.
+      // The unroll limit will be defined on the base name, if at all.
+      val baseName = IncrementalRenaming.getBase(name)
       // Read the unroll limit
-      val unrollLimit = getUnrollLimit(name, bodyMap).getOrThrow
+      val unrollLimit = getUnrollLimit(baseName, bodyMap).getOrThrow
       // Defines inlining k-times
       val kStepInline = inliner.kStepInline(unrollLimit, renaming)
       // Unroll all LET-IN definitions afterward
       val unrolledLetIn = unrollLetIn(bodyMap)(body)
       val inlined = kStepInline(unrolledLetIn)
       // Any remaining applications are default-replaced
       val defaultReplaced = replaceWithDefaultsTr(inlined)
+      // must specifically set .isRecursive to false, so no .copy
       TlaOperDecl(name, fparams, defaultReplaced)
-    case d @ TlaOperDecl(name, fparams, body) => // d.isRecursive = false
-      // Even though `name` is not recursive, it still may define recursive LET-IN operators inside
+    case d @ TlaOperDecl(_, _, body) => // d.isRecursive = false
+      // Even though the operator is not recursive, it still may define recursive LET-IN operators inside
       val unrolledLetIn = unrollLetIn(bodyMap)(body)
       if (body == unrolledLetIn) d
-      else TlaOperDecl(name, fparams, unrolledLetIn)
+      else d.copy(body = unrolledLetIn) // explicit construction is fine here, but .copy for consistency
     case _ => decl
   }
 
+  /**
+   * Finds the names of all recursive operators defined in LET-IN subexpressions of `ex`
+   */
+  private def allRecursiveLetInOperatorNames(ex: TlaEx): Set[String] = ex match {
+    case OperEx(_, args @ _*) =>
+      args.map(allRecursiveLetInOperatorNames).foldLeft(Set.empty[String]) { _ ++ _ }
+    case LetInEx(body, defs @ _*) =>
+      val allInDefs = defs.map(allRecursiveOperatorsInDecl).foldLeft(Set.empty[String]) { _ ++ _ }
+      val allInBody = allRecursiveLetInOperatorNames(body)
+      allInBody ++ allInDefs
+    case _ => Set.empty[String]
+  }
+
+  /**
+   * Finds the names of all recursive operators defined under decl.
+   * This includes recursiveoperators defined in LET-IN subexpressions,
+   * as well as the name of `decl` itself, if it is recursive.
+   */
+  private def allRecursiveOperatorsInDecl(decl: TlaOperDecl): Set[String] = {
+    // If this declaration is recursive, we take its base name (for LET-IN defined)
+    val selfOpt = if (decl.isRecursive) Some(IncrementalRenaming.getBase(decl.name)) else None
+    allRecursiveLetInOperatorNames(decl.body) ++ selfOpt
+  }
```

Comment:
Does it mean that we do not rely on SANY on identifying recursive operators?

Reply:
We do, but LET-IN defined recursive operators are not toplevel, so you need to scan the entire syntax tree for them.